Sources/Basic/POSIX.swift

```diff
@@ -0,0 +1,85 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright 2015 - 2016 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+// This file contains wrapper around POSIX functions to make them easier to use in swift.
+import libc
+import POSIX
+
+/// Extensions to the libc `stat` structure to interpret the contents in more readable ways.
+private extension libc.stat {
+
+     /// File system entity kind.
+     private enum Kind {
+         case file, directory, symlink, fifo, blockdev, chardev, socket, unknown
+
+         private init?(mode: mode_t) {
+             switch mode {
+                 case S_IFREG:  self = .file
+                 case S_IFDIR:  self = .directory
+                 case S_IFLNK:  self = .symlink
+                 case S_IFBLK:  self = .blockdev
+                 case S_IFCHR:  self = .chardev
+                 case S_IFSOCK: self = .socket
+             default:
+                 return nil
+             }
+         }
+     }
+
+     /// Kind of file system entity.
+     private var kind: Kind? {
+         return Kind(mode: st_mode & S_IFMT)
+     }
+ }
+
+/// Returns a structure containing information about the file system entity at `path`, or nil
+/// if that path doesn't exist in the file system.  Read, write or execute permission of the 
+/// file system entity at `path` itself is not required, but all ancestor directories must be searchable.
+/// If they are not, or if any other file system error occurs, this function throws a SystemError.
+/// If `followSymlink` is true and the file system entity at `path` is a symbolic link, it is traversed;
+/// otherwise it is not (any symbolic links in path components other than the last one are always traversed).
+/// If symbolic links are followed and the file system entity at `path` is a symbolic link that points to a
+/// non-existent path, then this function returns nil.
+private func stat(_ path: AbsolutePath, followSymlink: Bool = true) throws -> libc.stat? {
+    var sbuf = libc.stat()
+    let rv = followSymlink ? stat(path.asString, &sbuf) : lstat(path.asString, &sbuf)
+    guard rv == 0 || errno == ENOENT else { throw SystemError.stat(errno, path.asString) }
```

Comment:
My personal inclination currently is to first sort this out at the client level.

Currently `exists` doesn't throw, and I would prefer if we first migrate everything onto `FileSystem` and then figure out if it should throw. But of course that is also because I have a strong preference that it not throw.

I think making `POSIX.stat` itself throw makes sense.